dart/lib/src/reference.dart

```diff
@@ -0,0 +1,416 @@
+      var index = key;
+      if(index >= length) {
+        return null;
+      }
+      var elementOffset = _indirect + index * _byteWidth;
+      var flx = Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), 0);
+      flx._byteWidth = 1;
+      if (ValueTypeUtils.isTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.typedVectorElementType(_valueType);
+        return flx;
+      }
+      if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.fixedTypedVectorElementType(_valueType);
+        return flx;
+      }
+      var packedType = _buffer.getUint8(_indirect + length * _byteWidth + index);
+      return Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), packedType);
+    }
+    if (key is String && _valueType == ValueType.Map) {
+      var index = _keyIndex(key);
+      if (index != null) {
+        return _valueForIndex(index);
+      }
+    }
+    return null;
+  }
+
+  /// Get an iterable if the underlying flexBuffer value is a vector.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get vectorIterable {
+    if(isVector == false) {
+      throw Exception('Value is not a vector. It is: ${_valueType}');
+    }
+    return _VectorIterator(this);
+  }
+
+  /// Get an iterable for keys if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<String> get mapKeyIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapKeyIterator(this);
+  }
+
+  /// Get an iterable for values if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get mapValueIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapValueIterator(this);
+  }
+
+  /// Returns the length of the the underlying FlexBuffer value.
+  /// If the underlying value is [null] the length is 0.
+  /// If the underlying value is a number, or a bool, the length is 1.
+  /// If the underlying value is a vector, or map, the length reflects number of elements / element pairs.
+  /// If the values is a string or a blob, the length reflects a number of bytes the value occupies (strings are encoded in utf8 format).
+  int get length {
+    if (_length != null) {
+      return _length;
+    }
+    // needs to be checked before more generic isAVector
+    if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+      _length = ValueTypeUtils.fixedTypedVectorElementSize(_valueType);
+    } else if(_valueType == ValueType.Blob || ValueTypeUtils.isAVector(_valueType) || _valueType == ValueType.Map){
+      _length = _readInt(_indirect - _byteWidth, BitWidthUtil.fromByteWidth(_byteWidth));
+    } else if (_valueType == ValueType.Null) {
+      _length = 0;
+    } else if (_valueType == ValueType.String) {
+      var indirect = _indirect;
```

Comment:
Why are we copying this?  Seems unnecessary, nothing modifies `_indirect` in here.